include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h

```diff
@@ -34,181 +34,258 @@ namespace __par_backend_hetero
+
+        __reduce_event = __exec.queue().submit([&, __n, __n_items](sycl::handler& __cgh) {
+            __cgh.depends_on(__reduce_event);
+
+            sycl::accessor __temp_acc{__temp, __cgh, sycl::read_only};
+            sycl::accessor __res_acc{__res, __cgh, sycl::write_only, __dpl_sycl::__no_init{}};
+            __dpl_sycl::__local_accessor<_Tp> __temp_local(sycl::range<1>(__work_group_size2), __cgh);
+
+            __cgh.parallel_for<_LeafName...>(
+                sycl::nd_range<1>(sycl::range<1>(__work_group_size2), sycl::range<1>(__work_group_size2)),
                 [=](sycl::nd_item<1> __item_id) {
-                    ::std::size_t __global_idx = __item_id.get_global_id(0);
-                    ::std::uint16_t __local_idx = __item_id.get_local_id(0);
-                    // 1. Initialization (transform part). Fill local memory
-                    __transform_pattern(__local_idx, __n, __iters_per_work_item, __global_idx, /*global_offset*/ 0,
-                                        __temp_local, __rngs...);
-                    __dpl_sycl::__group_barrier(__item_id);
-                    // 2. Reduce within work group using local memory
-                    _Tp __result = __reduce_pattern(__item_id, __global_idx, __n_items, __temp_local);
-                    if (__local_idx == 0)
-                    {
-                        __reduce_pattern.apply_init(__init, __result);
-                        __res_acc[0] = __result;
-                    }
+                    __work_group_reduce_kernel<_Tp>(__item_id, __n, __n_items, __transform_pattern2, __reduce_pattern,
+                                                    __init, __temp_local, __res_acc, __temp_acc);
                 });
         });
 
         return __future(__reduce_event, __res);
     }
-}; // struct __parallel_transform_reduce_small_submitter
+}; // struct __parallel_transform_reduce_mid_submitter
 
-template <::std::uint16_t __work_group_size, ::std::size_t __iters_per_work_item, typename _Tp, typename _ReduceOp,
-          typename _TransformOp, typename _ExecutionPolicy, typename _Size, typename _InitType,
+template <::std::uint16_t __work_group_size, ::std::uint8_t __iters_per_work_item1,
+          ::std::uint8_t __iters_per_work_item2, typename _Tp, typename _ReduceOp, typename _TransformOp,
+          typename _ExecutionPolicy, typename _Size, typename _InitType,
           oneapi::dpl::__internal::__enable_if_device_execution_policy<_ExecutionPolicy, int> = 0, typename... _Ranges>
 auto
-__parallel_transform_reduce_small_impl(_ExecutionPolicy&& __exec, _Size __n, _ReduceOp __reduce_op,
-                                       _TransformOp __transform_op, _InitType __init, _Ranges&&... __rngs)
+__parallel_transform_reduce_mid_impl(_ExecutionPolicy&& __exec, _Size __n, const _ReduceOp& __reduce_op,
+                                     const _TransformOp& __transform_op, const _InitType& __init, _Ranges&&... __rngs)
```

Comment:
Just curious why Callables are now passed by `const` reference. Why is it better (keeping in mind the in the APIs all user callables come by value)?